### PR TITLE
fix(web, android, iOS): OSK 'tab' support when embedded

### PR DESCRIPTION
**Pull request by jahorton**
Fixes an error reported here: https://community.software.sil.org/t/tab-key-on-touch-layout/4291

When KeymanWeb is run in embedded mode, TAB should no longer function as a DOM-oriented command.  These changes allow us to disable that logic path when embedded, enabling `'\t'` output instead.

In-app appearance after a single press of TAB:

Screen Shot 2021-01-27 at 1 12 33 PM

I don't believe it's possible for iOS to support TAB key commands - Apple's pretty strict with what it allows keyboards to do.  While the [official documentation](https://developer.apple.com/library/archive/documentation/General/Conceptual/ExtensibilityPG/CustomKeyboard.html) doesn't explicitly state anything about tab commands, [this StackOverflow post](https://stackoverflow.com/a/28637083) states it to be impossible.

Compatibility of `\t` with predictive text... is probably best left to another time.  (Note:  those aren't the 'standard, default' corrections.)  Feels like an edge case...

Package used to test this:  [touch_tabbing.kmp.zip](https://github.com/keymanapp/keyman/files/5878032/touch_tabbing.kmp.zip)
- Github doesn't allow attaching .kmp files directly.

### CODE REVIEW

**Comment by mcdurdin** [comment]
I'd like to understand what action the Tab key on a physical bluetooth keyboard does on both iOS and Android before I approve this. What does Shift+Tab do?

I can try and test here with my Bluetooth keyboard.

**Comment by jahorton** [comment]
> I'd like to understand what action the Tab key on a physical bluetooth keyboard does on both iOS and Android before I approve this. What does Shift+Tab do?
> 
> I can try and test here with my Bluetooth keyboard.

The last PR for Android I see in this direction is https://github.com/keymanapp/keyman/pull/3664.  There's a fair shot that these changes would affect it, too, so it'd probably be good for @darcywong00 to chip in here.

**Comment by darcywong00** [comment]
> I'd like to understand what action the Tab key on a physical bluetooth keyboard does on both iOS and Android before I approve this. What does Shift+Tab do?

e.g. Google Keep
Tab on the external keyboard would navigate fields
Shift-Tab would navigate backwards

-----------------

I believe the galaxie_hebrew_positional keyboard uses K_TAB

```
+ "m" > U+05DE
U+05DE + [K_TAB] > U+05DD use(outkey)
```
So <kbd>m</kbd><kbd>TAB</kbd> should do something...

Though on this branch, typing in the Keyman app yields `U+05DE \t` (rtl)

**Comment by mcdurdin** [comment]
Okay, so I have tested on both my Android 10 phone and an iPhone 7.

Tab is supported on the hardware keyboard in both cases. Not terribly well by most apps, but certainly by some.

## Android

1. Tab moves forward a field in GMail
2. Shift+Tab moves back a field.
3. In some documents, Tab inserts a tab character.

## iPhone

1. Tab moves forward a field in GMail
2. Shift+Tab inserts a tab character
3. Nothing reliably moved back a field

So I think we need to treat this more like `K_ENTER` and pass the key event on to the app for it to decide what to do with, so long as it has not been overridden by the current Keyman keyboard, as @darcywong00 notes.

**Comment by jahorton** [comment]
> I believe the galaxie_hebrew_positional keyboard uses K_TAB
> 
> ```
> + "m" > U+05DE
> U+05DE + [K_TAB] > U+05DD use(outkey)
> ```
> 
> So mTAB should do something...
> 
> Though on this branch, typing in the Keyman app yields `U+05DE \t` (rtl)
> On `beta`, mTAB yields just `U+05DE`

That keyboard's OSK on iOS:

Screen Shot 2021-01-28 at 8 28 29 AM

I presume you were talking about hardware keystrokes on Android, then?  It wasn't particularly clear.

Interestingly, base KMW yields U+05DD in 'native'-mode KMW with said keyboard (desktop form-factor).

**Comment by jahorton** [comment]
... well, I figured out why that issue's appearing for that keyboard.  It's a pretty weird scenario.

```
U+05DE + [K_TAB] > U+05DD use(outkey)
U+05DE + [K_ENTER] > U+05DD use(outkey)

c ...


group(outkey) using keys

c empty group to trigger output of TAB and ENTER in KM9
```

The lines above translate into the following JS:

```
this.g0 = function(t, e) {
        var k = KeymanWeb
          , r = 0
          , m = 0;
        if (k.KKM(e, 16384, 9) && k.KFCM(1, t, ['מ'])) {
            r = m = 1;
            k.KDC(1, t);
            k.KO(-1, t, "ם");
            r = this.g1(t, e);
        }
        // ...
      }
      // ...
    }

    this.g1 = function(t, e) {
        var k = KeymanWeb
          , r = 0
          , m = 0;
        return r;
    }
```

Note something very important here:  the keyboard's group function is returning that _there is no match_, as a hack.  The idea behind the hack:  

```
c empty group to trigger output of TAB and ENTER in KM9
```

By saying "we didn't match anything", not only does the rule end up being applied in base Keyman, but the default behavior for those two keys kicks in as well.  This results in those keys being emitted by default, since "no rule matched".  (Well... except for on Web, where that TAB may carry a different meaning.  Side point - it didn't happen in my prior post b/c incomplete fix.)

The issue:  when embedded, KMW relies upon the whole "did a rule match?" thing for reporting the behavior of keystrokes, generating a "rule behavior" object for what the keyboard did.  But, since "no rule matched", the keyboard must not have done anything - so there's nothing to report to the host app.

I'd strongly argue that the bug is with the keyboard, not KMW.

**Comment by mcdurdin** [comment]
@jahorton writes:
> Note something very important here: the keyboard's group function is returning that _there is no match_, as a hack.

Okay, years ago this could have been called a hack. But it has been codified for many years now as core functionality in both Keyman for Windows and KeymanWeb. Quoting https://help.keyman.com/developer/language/guide/groups (which is identical to the same section in the [Keyman Developer *8.0* guide](https://help.keyman.com/developer/8.0/docs/guide_lang_groups)):

> If the final group processed is a context and keystroke group (using keys), and there is no nomatch rule, and the keystroke is not matched in the group, the keystroke will be output to the screen, regardless of whether or not it was matched in earlier groups.

@jahorton writes:
> I'd strongly argue that the bug is with the keyboard, not KMW. If not the keyboard, then the bug is with Developer's Web compiler, as I'm fairly certain the hack "works as intended" on Keyman for Windows.

The compiled keyboard code looks correct from my point of view.

It looks like we don't have any test cases exercising this, and perhaps the specification was a bit buried in the documentation, but it *did* work in earlier versions. I don't know if it ever worked for the embedded apps, though?

> The issue: when embedded, KMW relies upon the whole "did a rule match?" thing for reporting the behavior of keystrokes, generating a "rule behavior" object for what the keyboard did. But, since "no rule matched", the keyboard must not have done anything - so there's nothing to report to the host app.

This is a good analysis. And I agree that that sounds to me like the core issue. Sounds like we'll need to rework that in order to be 100% compatible with Keyman for Windows behaviour -- the return value should only be used to determine if the final keystroke should be passed through or not (and that is subject to nuance also as we may synthesize the output for normal character keys).

**Comment by jahorton** [comment]
> @jahorton writes:
> 
> > Note something very important here: the keyboard's group function is returning that _there is no match_, as a hack.
> 
> Okay, years ago this could have been called a hack. But it has been codified for many years now as core functionality in both Keyman for Windows and KeymanWeb. Quoting https://help.keyman.com/developer/language/guide/groups (which is identical to the same section in the [Keyman Developer _8.0_ guide](https://help.keyman.com/developer/8.0/docs/guide_lang_groups)):
> 
> > If the final group processed is a context and keystroke group (using keys), and there is no nomatch rule, and the keystroke is not matched in the group, the keystroke will be output to the screen, regardless of whether or not it was matched in earlier groups.

I'd argue that there's a difference between what the documentation says and your interpretation of it.  Granted, my nitpicking wouldn't have made much difference before predictive text support started coming into play.  Predictive text cares about the rule behaviors _themselves_, rather than just the resulting effects.

There's quite a difference in saying "no rules matched" vs "treat the key as if unmatched."  But, as you say:

> Sounds like we'll need to rework that in order to be 100% compatible with Keyman for Windows behaviour -- the return value should only be used to determine if the final keystroke should be passed through or not (and that is subject to nuance also as we may synthesize the output for normal character keys).

I _think_ I can make this work without too much issue.  I'll see if I can have something ready, including a test case, by the day's end if not sooner.

**Comment by mcdurdin** [comment]
> the keystroke will be output to the screen

I wince now when I read this. Even though I wrote it.

**Comment by jahorton** [comment]
I've reworked things according to our discussion above, and while doing so, I realized I could implement things without the need for Web-core to be aware of 'embedded' vs 'native' mode stuff.  So I fixed that up - Core really shouldn't need to know about that.

Still need to make a unit test, but the main fix is there.  I've also tested within the Android app, which now correctly produces the altered key 'followed' (b/c RTL) by `\t` output.

Screen Shot 2021-01-28 at 12 18 00 PM

**Comment by darcywong00** [comment]
I'm able to reproduce your latest output (including `\t`). 

Intermittently, I'm getting two characters when hitting an external keyboard key once.

Even with sil_euro_latin keyboard and clear context:
<kbd>l</kbd><kbd>TAB</kbd><kbd>l</kbd>, the textfield shows:
`l` `\t` `l` `l`

**Comment by jahorton** [comment]
> I'm able to reproduce your latest output (including `\t`).
> 
> Intermittently, I'm getting two characters when hitting an external keyboard key once.
> 
> Even with sil_euro_latin keyboard and clear context:
> lTABl, the textfield shows:
> `l` `\t` `l` `l`

Only on this branch, I take it?

**Comment by jahorton** [comment]
Well... I ended up using the ENTER key combos for the unit tests.  

With TAB, the headless test was easy enough, but the integrated, DOM-aware test had issues.  Default behavior for 'desktop' with TAB is to change the focused input element, which requires significant extra setup... and I think that synthetic events (used for simulation) won't actually trigger default behavior.

ENTER gets around that swimmingly, especially when using a `textarea` element; the consistency works in our favor for automated unit tests, and emulating use of the OSK key means that our code will always synthetically fill in the default, rather than relying on a browser to fill it in.

**Comment by jahorton** [comment]
> > I'm able to reproduce your latest output (including `\t`).
> > Intermittently, I'm getting two characters when hitting an external keyboard key once.
> > Even with sil_euro_latin keyboard and clear context:
> > lTABl, the textfield shows:
> > `l` `\t` `l` `l`
> 
> Only on this branch, I take it?
> 
> Also, how frequently? I don't see it at all when testing in 'native' web on the testing page, though I haven't gone super-rigorous with my tests for this yet.

I'm also not seeing it via Android Studio emulation, at least not in-app.  I only see the expected number of `l`s.  The closest thing to an issue I see is how short `\t` often is.

**Comment by darcywong00** [comment]
I'm using a bluetooth keyboard cause Android Studio doesn't really handle the tab key.
`beta` doesn't exhibit the "doubling" of characters

**Comment by jahorton** [comment]
How frequently do you see the 'doubling' effect?  Also, is there any form of regularity you can see?  (Such as... the key immediately after a TAB is always doubled?)

**Comment by darcywong00** [comment]
> doesn't really handle the tab key

Correction - other modifiers like <kbd>Ctrl</kbd> don't work with the emulator.

The doubling is maybe 1/3 times? Possibly after a TAB key...

File: common/core/web/input-processor/src/text/inputProcessor.ts [hunk 1]
```diff
@@ -180,7 +180,10 @@ namespace com.keyman.text {
       //        once THOSE are properly relocated.  (They're too DOM-heavy to remain in web-core.)
 
       // Only return true (for the eventual event handler's return value) if we didn't match a rule.
-      return ruleBehavior;
+
+      // Alas, the above notes are the goal, and will be separately addressed by #3695.  For now, a hacky
+      // workaround in order to not break default behaviors.
+      return ruleBehavior && !ruleBehavior.triggerKeyDefault;  // return `true` if the key is considered fully handled.
```
Comment (darcywong00) [inline common/core/web/input-processor/src/text/inputProcessor.ts:186]: I suspect this workaround is resulting in the intermittent behavior:
1. <kbd>TAB</kbd> from the external keyboard
2. KMW generates `\t` and this returns true
3. KMEA gets the status (true), and thinks KMW didn't process the key, so KMEA dispatches another Tab key to get handled.

keyboard.html (ll. 331-334)
```
         if (kmw['executeHardwareKeystroke'](code, shift, lstates)) { // false if matched, true if not
          // KMW didn't process the key, so have the Android app dispatch the key with the original event modifiers
          window.jsInterface.dispatchKey(code, eventModifiers);
        }
```

Comment (darcywong00) [reply common/core/web/input-processor/src/text/inputProcessor.ts:186]: I added a console.warn and can confirm with the Galaxie Hebrew keyboard
<kbd>m</kbd><kbd>TAB</kbd> generates the transformed character along with an extra `\t`.

Comment (jahorton) [reply common/core/web/input-processor/src/text/inputProcessor.ts:186]: > I suspect this workaround is resulting in the intermittent behavior:
> 
>     1. TAB from the external keyboard
> 
>     2. KMW generates `\t` and this returns true
> 
>     3. KMEA gets the status (true), and thinks KMW didn't process the key, so KMEA dispatches another Tab key to get handled.
> 
> 
> keyboard.html (ll. 331-334)
> 
> ```
>          if (kmw['executeHardwareKeystroke'](code, shift, lstates)) { // false if matched, true if not
>           // KMW didn't process the key, so have the Android app dispatch the key with the original event modifiers
>           window.jsInterface.dispatchKey(code, eventModifiers);
>         }
> ```

That sounds like what I expected with this:

> How frequently do you see the 'doubling' effect? Also, is there any form of regularity you can see? (Such as... the key immediately after a TAB is always doubled?)

KMW hadn't been processing it properly but now is... but there's then the matter of how to pass this along to Android.  Especially for cases where a command-style TAB (for change of focus) is expected.

Comment (darcywong00) [reply common/core/web/input-processor/src/text/inputProcessor.ts:186]: I think from #3664 we said for KMW to **not** process K_TAB or K_ENTER and let the embedded devices handle it.

Android would then interpret Tab depending on the field.

Comment (jahorton) [reply common/core/web/input-processor/src/text/inputProcessor.ts:186]: > I think from #3664 we said for KMW to **not** process K_TAB or K_ENTER and let the embedded devices handle it.
> 
> Android would then interpret Tab depending on the field.

Just to be clear... just for hardware-based keystrokes, right?  OSK-based keystrokes require KMW emulation/handling, after all.  Or is there a way to make an OSK-based keystroke _act_ like a hardware keystroke on Android?

Comment (darcywong00) [reply common/core/web/input-processor/src/text/inputProcessor.ts:186]: So for a keyboard like galaxie_hebrew_positional, KMW should match K_TAB and return `false` to KMEA.
For sil_euro_latin, KMW doesn't match K_TAB and should return `true` to KMEA to handle the tab.

Comment (jahorton) [reply common/core/web/input-processor/src/text/inputProcessor.ts:186]: > I think from #3664 we said for KMW to **not** process K_TAB or K_ENTER and let the embedded devices handle it.
> 
> Android would then interpret Tab depending on the field.

> So for a keyboard like galaxie_hebrew_positional, KMW should match K_TAB and return `false` to KMEA.

Pick one, please.  The first quote above indicates that I should return `true` to KMEA.  Hence my question above:




> > I think from #3664 we said for KMW to **not** process K_TAB or K_ENTER and let the embedded devices handle it.
> > Android would then interpret Tab depending on the field.
> 
> Just to be clear... just for hardware-based keystrokes, right? OSK-based keystrokes require KMW emulation/handling, after all. Or is there a way to make an OSK-based keystroke _act_ like a hardware keystroke on Android?

Comment (jahorton) [reply common/core/web/input-processor/src/text/inputProcessor.ts:186]: @marc - My interpretation of all of this:

For keystrokes that match _a_ rule, but not within the final processing group:

- When embedded and the keystroke was received from a physical keyboard:
    - KMW should NOT emit the keystroke's default behavior
    - KMW should indicate that it was unhandled, thus allowing KMEA to do it.
    - KMEA should then update the text-context accordingly, allowing predictive text to fill it in.
- When the keystroke was received from the OSK:
    - KMW should emit the keystroke's default behavior.
        - Note:  when in 'native', within a browser this can trigger focus changes.
        - Is that _possible_ to do from 'synthetic' keystrokes for Android?  Or do we just give up there?
- When not embedded and the keystroke was received from a physical keyboard:
    - KMW should not (directly) emit the keystroke's default behavior
    - The key-handling event handlers indicate that the event wasn't handled, triggering the browser to do perform default behaviors
    - Predictive text, if active, should note any context changes triggered as a result (NOTE:  not currently done!) and update predictions accordingly.

My issues with this breakdown so far:
- I don't think this will (currently) be handled well by Android default-handling rules.
    - What if a plain-text key had a similar setup?
    - Would our Android app properly emit that 'a'?  My impression is that it has special handling for `K_TAB` and `K_ENTER`, but I could be wrong.
- This markedly complicates predictive text.
    - For `galaxie_hebrew_positional`, this behavior only exists for `K_TAB` and `K_ENTER`... so we kind of luck out re: predictive text.
    - Suppose a key's default behavior is to emit 'a' and has a matching rule (but not in the final match group)... and wasn't pressed.  How would predictive text know to include that 'a'?
    - My current approach:  for possible 'alternate' keystrokes, follow 'synthetic'/OSK-based rules:  always model text-based default behaviors (except for TAB and ENTER in 'native' mode)
    - The issue is then that the original keystroke (from hardware) doesn't come from the same processing path as the fat-finger alternates (from synthetic / OSK processing).
    - There's _also_ the issue of recognizing _after the fact_ that the emitted default behavior (example: 'a') is actually a part of the previously-processed keystroke, rather than treating it as a new, separate 'keystroke' upon the (delayed) context update.

Comment (jahorton) [reply common/core/web/input-processor/src/text/inputProcessor.ts:186]: For @darcywong00:

> Android would then interpret Tab depending on the field.

_When_ does Android's OS allow us to do this?  Can we do this at any time with the correct setup?

_What_ is required, by Android's APIs, to inform the Android OS that this should be done?

**Comment by jahorton** [comment]
Due to the rising complexity noted during our team discussions, we're postponing this until 15.0.  We're too late in the Beta process to address this properly right now.

**Comment by mcdurdin** [comment]
Lots to think about:
* Enter key action
* Tab key action
* Backspace behaviour in text and at start of context.

A reference: 
https://developer.android.com/reference/android/view/inputmethod/EditorInfo#IME_MASK_ACTION

**Comment by jahorton** [comment]
Since we're postponing this past 14.0, I've gone ahead and created two new issues for tracking the underlying problems:

Fixes https://github.com/keymanapp/keyman/issues/4498.
Fixes https://github.com/keymanapp/keyman/issues/4499.

**Comment by jahorton** [comment]
Note:  only _some_ of these changes were actually merged; I manually split out the TAB-oriented parts for #4498, so related changes seen in this PR did _not_ get merged in.  Apparently, the way I did it managed to confuse GitHub - glad I noticed the issue quickly.
